proto/public/uber/cadence/api/v1/service_domain.proto

```diff
@@ -0,0 +1,129 @@
+// THE SOFTWARE.
+
+syntax = "proto3";
+
+package uber.cadence.api.v1;
+
+option go_package = "apiv1";
+option java_multiple_files = true;
+option java_outer_classname = "ApiProto";
+option java_package = "com.uber.cadence.api.v1";
+
+import "google/protobuf/duration.proto";
+import "dependencies/gogoproto.proto";
+import "uber/cadence/api/v1/domain.proto";
+
+service DomainAPI {
+  // RegisterDomain creates a new domain which can be used as a container for all resources.  Domain is a top level
+  // entity within Cadence, used as a container for all resources like workflow executions, task lists, etc.  Domain
+  // acts as a sandbox and provides isolation for all resources within the domain.  All resources belongs to exactly one
+  // domain.
+  rpc RegisterDomain(RegisterDomainRequest) returns (RegisterDomainResponse);
+
+  // DescribeDomain returns the information and configuration for a registered domain.
+  rpc DescribeDomain(DescribeDomainRequest) returns (DescribeDomainResponse);
+
+  // ListDomains returns the information and configuration for all domains.
+  rpc ListDomains(ListDomainsRequest) returns (ListDomainsResponse);
+
+  // UpdateDomain is used to update the information and configuration for a registered domain.
+  rpc UpdateDomain(UpdateDomainRequest) returns (UpdateDomainResponse);
+
+  // DeprecateDomain us used to update status of a registered domain to DEPRECATED.  Once the domain is deprecated
+  // it cannot be used to start new workflow executions.  Existing workflow executions will continue to run on
+  // deprecated domains.
+  rpc DeprecateDomain(DeprecateDomainRequest) returns (DeprecateDomainResponse);
+}
+
+message RegisterDomainRequest {
+  string name = 1;
+  string description = 2;
+  string owner_email = 3;
+  google.protobuf.Duration workflow_execution_retention_period = 4 [(gogoproto.stdduration) = true];
+  bool emit_metric = 5;
+  repeated ClusterReplicationConfiguration clusters = 6;
+  string active_cluster_name = 7;
+  map<string, string> data = 8;
+  string security_token = 9;
+  bool is_global_domain = 10;
+  ArchivalStatus history_archival_status = 11;
+  string history_archival_uri = 12;
+  ArchivalStatus visibility_archival_status = 13;
+  string visibility_archival_uri = 14;
+}
+
+message RegisterDomainResponse {
+}
+
+message DescribeDomainRequest {
+  string name = 1;
+  string id = 2;
+}
+
+message DescribeDomainResponse {
+  DomainInfo info = 1;
+  DomainConfiguration configuration = 2;
+  DomainReplicationConfiguration replication_configuration = 3;
+  int64 failover_version = 4;
+  bool is_global_domain = 5;
+}
+
+message ListDomainsRequest {
+  int32 page_size = 1;
+  bytes next_page_token = 2;
+}
+
+message ListDomainsResponse {
+  repeated DescribeDomainResponse domains = 1;
+  bytes next_page_token = 2;
+}
+
+message UpdateDomainInfo {
+  string description = 1;
+  string owner_email = 2;
+  map<string, string> data = 3;
+}
```

Comment:
only used in one place (below), and the others have these just flat in the parent message.  maybe worth flattening into the update request?  though that will have the semantic change of no longer being nullable.

Reply:
Not sure about this. It kind of holds intent to update these fields. If we flatten them, then these fields come together with `name`. This might imply that we want to update name as well, rather than use it a domain selection field.

Reply:
I'm guessing pretty much just "name" isn't modifiable in the update request?  or are there others?

also, something I'm not familiar with: are _all_ of the fields (that are modifiable, including in nested messages) required to be non-empty?  otherwise we may not be able to tell "remove X" from "do not modify X"

Reply:
- `DomainConfiguration.emit_metric` is a bool, so if the configuration is provided it'll have to be set in the request, or it'll be disabled.
- `*_archival_uri` are just strings, so `""` is ambiguous between "leave alone" and "remove", but maybe that's fine in practice (e.g. if status is invalid assume "no change", require explicit disable to delete)

otherwise they _appear_ pretty zero-value friendly to me.

Reply:
At least in CLI (I think currently the only place), before issuing update request the current state is read and only changed fields are are overwritten. Thus all of them are provided in the update request.

However you made a good point and leaving a distinction for "do not modify" makes sense to me. In that case it makes sense to flatten all inner nested fields wrap each of them with a pointer.

Reply:
minor: imo we should drop DomainConfiguration.emit_metric. This option adds more confusion than usefulness. Practically I haven't see any real use case to require turning metrics off (and I am not sure how to operate the cluster if that's the case).
We can do it in a different PR though if that makes things easier.

Reply:
Dropped `emit_metric`